rfcs/framework-0063-notification-system.md

```diff
@@ -0,0 +1,422 @@
+        UserEmailBackend,
+    ]
+
+#
+# Payload
+#
+# invenio_requests/notifications.py
+class RequestPayload:
+    def __call__(self, request=None, **kwargs):
+        return {"request": request.to_dict()}
+
+# invenio_rdm_records/notifications.py
+class RecordPayload:
+    def __call__(self, record=None, **kwargs):
+        return {"record": record.to_dict()},
+
+# invenio_communities/notifications.py
+class CommunityPayload:
+    def __call__(self, community=None, **kwargs):
+        return {"community": community.to_dict()}
+
+#
+# Recipients
+#
+# invenio_requests/notifications.py
+class RequestRecipients:
+    def __init__(self, include_creator=True, include_receiver=True):
+        self.include_creator = include_creator
+        self.include_receiver = include_receiver
+
+    def __call__(self, notification, recipients: list):
+        ret = []
+        if self.include_creator:
+            ret.append(request.created_by.resolve())
+        if self.include_receiver:
+            ret.append(request.receiver.resolve())
+        return ret
+
+# invenio_communities/notifications.py
+class CommunityRecipients:
+    def __init__(self, roles=None):
+        self.roles = roles
+
+    def __call___(self, notification, recipients: list):
+        ret = []
+        for rec in recipients:
+            if isinstance(rec, Community):
+                comm = rec
+                members = Member.get_members(comm.id)
+                for m in members:
+                    if not m.user_id:
+                        continue
+                    if self.roles and m.role not in self.roles:
+                        continue
+                    user = m.relations.user.dereference()
+                    notif_pref = user.preferences["notifications"]
+                    com_pref = notif_pref.get(comm.id)
+                    if com_pref:
+                        # check if notification is enabled/disabled for the member
+                        # if ...:
+                        #     continue
+                    ret.append(user)
+            else:
+                ret.append(rec)
+        return ret
+
+# invenio_users_resources/notifications.py
+class UserRecipient:
+    def __call__(self, notification, recipients: list)
+        ret = []
+        for rec in recipients:
+            if isinstance(rec, User):
+                if rec.preferences["notifications"]["enabled"]
+                    ret.append(rec.dump())
+            else:
+                ret.append(rec)
+        return ret
+
+#
+# Backends
+#
+# invenio_users_resources/notifications.py
+class UserEmailBackend:
+    def __call__(self, notification, recipients: list)
+        for rec in recipients:
+            user = rec.data
+            rec.backends.append({
+                "backend": "email",
+                "to": f"{user.profile.full_name} <{user.email}>",
+            })
+
+#
+# Usage in a service method
+#
+uow.register(NotificationOp(CommunityRecordInclusionNotificationBuilder.build(
```

Comment:
Where would `NotificationOp` process things, `post_commit`? would it be worth to add it to the RFC?

Reply:
The `NotificationOp` is not doing any processing, but delegates to a `NotificationBuilder` creating the notification payload/context containing just entity references to the underlying objects. The "heavy" processing has been updated to take place in the `NotificationManager` class in a Celery task.